README.md
# Instalador Desatendido (Semi Automatizado) para OpenStack (JUNO)
Reynaldo R. Martinez P.
E-Mail: TigerLinux@Gmail.com
Caracas, Venezuela.

## Introducción

Este instalador fue realizado para automatizar las tareas de creación de una
infraestructura de virtualización basada en OpenStack. Hasta el momento,
existen cuatro "sabores" del instalador, uno para Debian 7, uno para Centos 7,
y finalmente uno para Ubuntu 14.04 LTS.

Las tres versiones producen un OpenStack utilizable en producción, pero recomen-
damos en el caso de Centos 7 esperar a que el sistema operativo madure lo sufi-
ciente antes de considerarlo realmente listo para producción.

En resumen, este instalador puede producir un OpenStack Juno completamente
utilizable en ambientes de producción, sin embargo, recuerde que el factor "bugs"
no depende de nosotros, sino del proyecto OpenStack y de los proyectos que 
empaquetan para centos, debian y ubuntu.

También tome en cuenta que "hasta la fecha" (y eso podría cambiar en el futuro)
los paquetes de Centos (provistos por el proyecto RDO de RedHat) tienden a
actualizarse mas a menudo que los paquetes para Debian 7 en "gplhost".

Los paquetes de Ubuntu son los que normalemte están mas actualizados.

## Uso del Instalador.

### Primero

*LEA, LEA, LEA y luego de descansar de leer, LEA DE NUEVO !!.*

Lea todo lo que puede de *OpenStack* si desea incursionar en el área de
virtualización en la nube.  Si no le gusta leer, entonces apóyese en alguien
que si tenga la disposición para hacerlo, pero no trate de usar este
instalador sin tener algún tipo de conocimientos a la mano. Vea el archivo
`NOTAS.txt` para entender un poco mas de cuales conocimientos usted debería
tener.

El sub-mundo de *OpenStack* engloba varias tecnologías del mundo de Software
Libre y del mundo de Redes que deben ser comprendidas muy bien antes de
siquiera intentar hacer cualquier instalación de openstack ya sea con esta
herramienta de instalación o con cualquier otra. En resumen, si usted no tiene
los conocimientos, no lo intente. Adquiera los conocimientos primero, y luego
proceda !.

Antes de usar el instalador, debe preparar su o sus servidores. De nuevo, en
el archivo `NOTAS.txt` hay puntos importantes que deben ser cubiertos antes de
iniciar una instalación usando estas herramientas. El instalador hará una
serie de validaciones que, en caso de arrojar resultados negativos, hará que
aborte el proceso.

### Segundo: Modifique el archivo de configuración del instalador.

El instalador tiene un archivo central de configuración:
`./configs/main-config.rc`. Dicho archivo está bastante documentado, de manera
que si usted hizo *su tarea* y estudió de *OpenStack*, sabrá que debe modificar
en el mismo. Hay cosas muy obvias como *contraseñas*, *direcciones IP*,
*módulos a instalar* y *dominios de dns*.

En su versión por defecto, el archivo de configuración tiene las selecciones
de módulos para instalar lo que se conoce como un "all-in-one" (un servidor
monolítico de *OpenStack* con todos los servicios).

Adicionalmente, existen tres módulos que por defecto están en "no":
*Heat*, *Swift* y *SNMP*. El módulo de swift se puede instalar "si usted
realmente va a usarlo". *Swift* por si solo ya es casi tan extenso como todo 
*OpenStack*. Úselo si **REALMENTE** sabe lo que está haciendo y si **REALMENTE**
lo va a utilizar. El módulo de *SNMP* instala variables de monitoreo útiles si
usted va a monitorear *OpenStack* vía *SNMP* pero no instala ningún tipo de
aplicación de monitoreo. Las variables están descritas (si usted instala el
soporte) en el archivo `/etc/snmp/snmpd.conf`.

NOTA: Se incluyen archivos para el agente ZABBIX en el directorio "Goodies".

Si usted desea instalar un "all-in-one", sólo modifique las contraseñas,
direcciones IP y dominios de correo y de *dhcp/dnsmasq* que aparecen en el
archivo de configuración.

Luego de actualizar su archivo de configuración, ejecute en la raíz del
directorio del *script* el comando siguiente:

```
# ./main-installer.sh install
```

El instalador le preguntará si quiere proceder (y/n).

Si ejecuta el instalador con el parámetro adicional *auto*, el mismo seguirá
de manera automática sin preguntar. Ejemplo:

```
# ./main-installer.sh install auto
```

Usted puede guardar todas las salidas del instalador usando la herramienta
`tee`. Ejemplo:

Si su nodo de controller va a incluir un servicio de compute (controller +
compute), la siguiente variable del archivo de configuración debe estar en
"no":

```bash
nova_without_compute="no"
```

En cambio, si va a instalar un controlador "puro" (sin servicio de compute)
coloque la variable en "yes":

```bash
nova_without_compute="yes"
```

**Nodos de compute:** Para los nodos de compute, debe dejar en "yes" sólo las
variables de instalación de los módulos de nova y neutron. El resto de los
módulos (glance, cinder, horizon, etc.) deben estar en "no".  Adicionalmente,
las siguiente variables en las secciones de nova y neutron deben estar en
"yes":

```bash
nova_in_compute_node="yes"
neutron_in_compute_node="yes"
```

Debe colocar las IP's de los servicios de neutron, keystone, glance y cinder
según la que tiene el controlador (incluyendo las Ip's del backend de Base de
Datos). En cambio, las siguientes variables deben ser colocadas a la IP del
nodo de compute:

```bash
novahost="IP del Nova Compute Host"
glancehost="IP del Controlador"
cinderhost="IP del controlador"
neutronhost="IP del controlador"
keystonehost="IP del controlador"
messagebrokerhost="IP del controlador"
dbbackendhost="IP del controlador o del backend de base de datos"
vncserver_controller_address | spiceserver_controller_address = "IP del controlador"
```

### Backend de Base de Datos

El instalador tiene la posibilidad de instalar y configurar el backend de base de datos, y de crear
las bases de datos. Esto es completamente controlable por el archivo de configuración a través de las
siguientes variables:

```bash
dbcreate="yes"
dbinstall="yes"
dbpopulate="yes"
```

Con estas tres opciones en "yes", se instalará el software de base de datos,
se configurará y se crearán las bases de datos, todo con la información
contenida en el archivo de configuración.

> **ALERTA**: Si usted elige estas opciones, se debe asegurar que no exista
> previamente software de base de datos instalado o el proceso fallará.

Si usted desea "no instalar el software" pero si tiene acceso administrativo
completo al backend de base de datos, coloque las siguiente combinación:

```bash
dbcreate="yes"
dbinstall="no"
dbpopulate="yes"
```

Con esto, el software de base de datos no será instalado, pero queda de parte
de usted (o su *DBA*) asegurarse que tiene acceso administrativo completo para
crear y modificar bases de datos en el backend seleccionado.

Si no desea ni instalar ni crear bases de datos (asumimos que ya las tiene
previamente creadas) coloque los tres valores en "no":

```bash
dbcreate="no"
dbinstall="no"
dbpopulate="no"

```

### Backend de Mensajería (Message Broker)

Como parte de los componentes a instalar y configurar, este instalador instala
y configura el software para *AMQP* (el *Message Broker*). Esto es un paso
mandatorio para un controlador o un *all-in-one*. Si su o sus servidores ya
tienen un messaje broker instalado, pueden ocurrir conflictos que prevengan el
correcto funcionamiento de la instalación.


### Administrador de Consola (NOVNC/SPICEHTML5)

Mediante una opción configurable en ./configs/main-config.rc (consoleflavor), usted
puede elegir que administrador de consola a usar (NoVNC o SpiceHTML5). Le sugerimos
usar NoVNC que da menos problemas con configuraciones de teclado no us-english y es
mas fácil de reconfigurar para SSL.


### Trove

Este instalador tiene la opción de instalar Trove (ya estable a partir de OS-Icehouse). El
módulo de instalación de Trove hace la configuración base de Trove pero NO CREA las imágenes
ni reconfigura los datastores para dichas imágenes. Eso le toca a usted como administrador
de la nube de OpenStack (jejeje).


### Scripts de Ayuda

Este instalador colocará en /usr/local/bin un script de ayuda para poder levantar, bajar,
desactivar, activar los servicios de openstack:

```bash
openstack-control.sh OPCIÓN
```

El script utiliza las siguientes opciones:

1. **enable**: activa los servicios para autoarranque en el boot del servidor.
2. **disable**: desactiva los servicios para autoarranque en el boot del servidor.
3. **start**: arranca todos los servicios.
4. **stop**: detiene todos los servicios.
5. **restart**: reinicia todos los servicios.
6. **status**: muestra el estado de todos los servicios.

    > **IMPORTANTE**: El script "openstack-control.sh" tiene la gran ventaja de subir (o
    > bajar) todos los servicios de openstack en el orden correcto. Tanto los
    > paquetes de debian/ubuntu como los de centos colocan el orden no precisamente
    > "óptimo". Recomendación: colocar lo siguiente en el /etc/rc.d/rc.local del
    > servidor para forzar a que los servicios de OpenStack arranquen en el orden
    > correcto:

```bash
/usr/local/bin/openstack-control.sh restart
```

> NOTA: El script detecta que servicios de OpenStack fueron instalados y
> configurados por el instalador para
> iniciar/detener/reiniciar/habilitar/deshabilitar/verificar los servicios
> correctos en el orden correcto.

Adicional a `openstack-control.sh`, el instalador copia a `/usr/local/bin` el
script `openstack-log-cleaner.sh`.  Este script tiene como función limpiar
todos los logs de todos los componentes de *OpenStack* instalados por esta
herramienta de instalación.

El script es llamado durante la fase final de instalación para limpiar los
logs antes de dejar el servidor instalado, pero puede ser utilizado también
para limpiar los logs en caso de ser necesario.

NOTA IMPORTANTE: Recomendamos usar el script openstack-control.sh para inicializar
todos sus servicios de openstack !. Coloque todos los servicios en "disable" con
"openstack-control.sh disable" y llame al script con la opción "start" desde
el /etc/rc.local del sistema operativo !.

### ARRANQUE CONTROLADO DE MAQUINAS VIRTUALES

Script "openstack-vm-boot-start.sh": Este script sirve para levantar las VMs
de manera controlada y con tiempos de espera entra cada inicio de VM con el fin
de evitar "tormentas" de I/O. El script utiliza el siguiente archivo de
configuración:

```bash
/etc/openstack-control-script-config/nova-start-vms.conf
```

Coloque en dicho archivo los nombres de las VMs que quiere arrancar de manera
ordenada y secuencial para evitar tormentas de I/O. El script puede ser llamado
desde el "/etc/rc.local" del sistema operativo.

NOTA: Los nombres de las VMs se deben obtener del comando "nova list".


### DNSMASQ

El instalador crea una configuración personalizada del componente *dnsmasq*
usado por el agente *DHCP* de *Neutron* (`neutron-dhcp-agent`). Dicha configuración
incluye un archivo donde usted puede colocar opciones especiales:

```
/etc/dnsmasq-neutron.d/neutron-dnsmasq-extra.conf
```

Hay ejemplos comentados en el archivo. Use esos ejemplos para pasar opciones a
las distintas instancias de dnsmasq creadas para cada subred donde usted
active la opción de usar *dhcp*.

> **Recomendación**: Trate de tener una buena estructura de "DNS" para sus máquinas
> virtuales de manera que las pueda identificar de manera apropiada y utilice
> las opciones de dnsmasq para agregar o separar subdominios para cada rango de
> direcciones IP.


### Modularización del Instalador

Si bien el proceso de instalación principal "*main-installer*" se encarga de
llamar a cada módulo de cada componente del instalador, dichos módulos son
realmente independientes entre si, al punto de que pueden ser llamados de
manera secuencial. No es el caso común, pero puede hacerse. El orden normal de
ejecución de cada módulo es el siguiente (asumiendo que todos los componentes
serán instalados):

* requeriments.sh
* messagebrokerinstall.sh
* databaseinstall.sh
* requeriments-extras.sh (sólo presente para Debian 7 y Ubuntu Server 14.04 LTS)
* keystoneinstall.sh
* swiftinstall.sh
* glanceinstall.sh
* cinderinstall.sh
* neutroninstall.sh
* novainstall.sh
* ceilometerinstall.sh
* heatinstall.sh
* troveinstall.sh
* snmpinstall.sh
* horizoninstall.sh
* postinstall.sh

Si desea ejecutar los módulos en orden sin usar el instalador principal, debe ejecutarlos desde el
directorio donde se encuentra el instalador, de la siguiente manera:

```
# ./modules/Modulo-a-ejecutar.sh
```

Todos los módulos leen el archivo de configuración
(`./configs/main-config.rc`) y tomarán las opciones de configuración desde
dicho archivo.

Cada módulo que se ejecuta de manera exitosa dejará una serie de archivos de
control en el directorio siguiente:

```
/etc/openstack-control-script-config
```

Dichos archivos de control son utilizados por los módulos para evitar una
re-ejecución de los mismos que pueda causar problemas en la instalación de
OpenStack, y también son utilizados por el script "openstack-control.sh" para
saber que componentes de OpenStack fueron instalados de manera exitosa y que
servicios están instalados para iniciar/detener/reiniciar/etc.


### RECOMENDACIONES DE INSTALACIÓN PARA CENTOS Y DEBIAN.

#### Centos 6:

1. Instale centos con la selección de paquetes para "Basic Server" (servidor
   básico). Asegúrese de tener correctamente instalado, configurado y operativo
   el servicio ntpd. Se recomienda también usar ntpdate.

2. Agregue los repositorios EPEL y RDO (ver "NOTAS.txt").

3. Instale y configure OpenVSWitch (de nuevo, ver "NOTAS.txt").

ALERTA: Esta versión de OpenStack no soporta MySQL menor a 5.5. Vea las notas
y tome sus precauciones !.

La versión para CENTOS 6 de este instalador usa MariaDB 5.5 disponible en los
repositorios RDO.

Si usted va a usar un manejador de base de datos externo basado en MySQL/MariaDB,
ASEGURESE que sea la versión correcta (5.5) para no terminar con una instalación
de OpenStack inservible !.


#### Centos 7:

1. Instale centos con la selección de paquetes para "Infraestructure Server" (ser-
   vidor de infraestructura). Asegúrese de tener correctamente instalado, configu-
   rado y operativo el servicio ntpd. Ser recomienda también usar ntpdate.

2. Agregue los repositorios EPEL y RDO (ver "NOTAS.txt").

3. Instale y configure OpenVSWitch (de nuevo, ver "NOTAS.txt").

ALERTA: Esta versión de OpenStack no soporta MySQL menor a 5.5. Vea las notas
y tome sus precauciones !.

Si usted va a usar un manejador de base de datos externo basado en MySQL/MariaDB,
ASEGURESE que sea la versión correcta (5.5) para no terminar con una instalación
de OpenStack inservible !.

NOTA IMPORTANTE: El instalador de Centos 7 desactiva SELINUX. Esto se requiere para
evitar problemas encontrados especialmente cuando se utiliza PostgreSQL como backend
de base de datos y con NOVA-API. Cuando estos bugs sean corregidos, se volverá a reactivar
SELINUX. Por lo pronto, lo desactivamos desde el módulo de requerimientos.


#### Debian 7:

1. Instale debian con la selección de paquetes "Standard System Utilities"
(utilitarios de sistema estandard) y con SSH Server (servidor ssh). Asegúrese
de tener correctamente instalado, configurado y operativo el servicio ntpd. Se
recomienda también usar ntpdate.

2. Agregue el repositorio de OpenStack para Debian y asegúrese de tener las ramas completas para sus repos
de debian (ver "NOTAS.txt").

3. Instale y configure OpenVSWitch (de nuevo, ver "NOTAS.txt").

#### Ubuntu 14.0.4 LTS:

1. Instale Ubuntu Server 14.04 LTS de manera estandar y seleccione como paquete adicional
"OpenSSH Server". Instale y configure el servicio ntpd. Se recomienda también usar ntpdate.

2. Instale y configure OpenVSWitch (ver "NOTAS.txt").

### Cinder:

Si va a usar CINDER con lvm-iscsi, asegúrese de tener una partición o disco libre para crear un LVM
llamado "cinder-volumes". Ejemplo (disco libre /dev/sdc):

```bash
pvcreate /dev/sdc
vgcreate cinder-volumes /dev/sdc
```

Otro ejemplo con una partición libre /dev/sda3:

```bash
pvcreate /dev/sda3
vgcreate cinder-volumes /dev/sda3
```

### Swift:

Si va a usar swift, recuerde que debe tener el disco/partición de swift montado sobre un directorio
específico y este debe ser indicado en la configuración del instalador (main-config.rc).

Ejemplo:

Variable `swiftdevice="d1"`

En el fstab "d1" debe estar montado así:

```
/dev/sdc1 /srv/node/d1 ext4 acl,user_xattr 0 0
```

En este ejemplo, se asume que existe ya una partición "/dev/sdc1" previamente formateada en ext4 o
cualquier otro sistema de archivos soportado por Linux.


### Arquitectura:

Ya sea que use Centos, Debian o Ubuntu, debe elegir utilizar 64 bits
(amd64/x86_64). No trate de instalar OpenStack en 32 bits.

### Servicio NTP:

Es VITAL que sus servidores de OpenStack tengan una correcta sincronización de
tiempo o habrá serios problemas entre el controller y los nodos de
compute. Lea la documentación de OpenStack para saber mas al respecto.


### Recomendaciones para Virtualbox.

Usted puede usar este instalador en una VM de VirtualBox si no desea usar un
servidor "real" para practicar y aprender OpenStack. Debería tener un "mínimo
absoluto" de 2GB's de RAM en su equipo "real" y asignar al menos 900 Mb's de
RAM a la VM de VirtualBox, pero si quiere hacer una prueba mas extensa, trate
de tener una VM con al menos 4 GB's.

### Recomendaciones de hardware para una VM de VirtualBox:

Discos Duros: Uno para el sistema operativo (16 Gb's mínimo), uno para
Cinder-Volumes y otro para swift - espacio variable... mínimo 8GB's para cada
disco (switf y cinder-volumes).  Red: tres interfaces:
 * Interfaz 1 en modo NAT para salida a Internet.
 * Interfaz 2 en modo "host only adapter, promisc: todos", para poder
   administrar el Servidor OpenStack - sugerencia: Usar vboxnet0 con la red
   192.168.56.0/24 (desactivar el dhcp de virtualbox) y asignarle la IP

DOCS/CHANGELOG.txt
1.0.8.el6 - 28-Febrero-2015
- Mejoras en módulo de instalación de ceilometer.

1.0.7.el6 - 16-Febrero-2015
- Se eliminan variables SNMP/ZABBIX para conteo de CPU y
  RAM usada por QEMU. Las variables han demostrado ser
  inefectivas. Cuando se encuentre otro método viable y
  rápido para obtener estas métricas, se incluirán de nuevo
  las variables.
- Se hacen mejoras en la secuencia de configuración de
  libvirt.

1.0.6.el7 - 19-Ene-2015
- Se corrige un bug en los parámetros de VNC y SPICE en el
  módulo de instalación de nova.
- Se agrega documentación sobre implementación en ambientes
  de VLAN's. Vea el archivo en el directorio DOCS.

1.0.5.el7 - 11-Ene-2015
- Se incluyen settings para acceso a BD en los módulos
  de keystone, glance, cinder, neutron, nova y heat.

1.0.4.el7 - 09-Dic-2014
- Se corrige un BUG en la configuración del parámetro
  "cpu_mode" en el módulo de instalación de Nova.

1.0.3.el7 - 26-Nov-2014
- Se incluye la opción de seleccionar el tipo
  de proveedor de token para Keystone (pki o
  uuid).
- Corrección menor en módulo de instalación de
  keystone.

1.0.2.el7 - 26-Oct-2014
- Mejoras para instalación de ceilometer en
  distintas configuraciones de nodos de con-
  trolador o compute.

1.0.1.el7 - 22-Oct-2014
- RELEASE!!!!. Primera versión de producción
  basada en paquetes ya oficiales en RDO.
- Correcciones menores para metricas de CPU,
  disco y red entre Nova-Computer y Ceilometer.

1.0.0.rc1.el7 - 19-Oct-2014
- Primer RC basado en los paquetes RC1/RC2/RC3
  en el repositorio RDO.
- Limpieza general de las configuraciones y
  adaptación de items "deprecados" a sus nuevas
  versiones en todos los módulos donde fue
  necesario hacer los cambios.

0.0.4.beta.el7 - 15-Oct-2014
- Se agrega autenticación y se refuerza la
  seguridad entre MongoDB y Ceilometer.
- Cambios adicionales para asegurar la
  correcta instalación y configuración de
  RabbitMQ.
- Cambios menores en secuencia de instalación
  de neutron respecto a aprovisionamiento de
  la base de datos.

0.0.3.beta.el7 - 14-Oct-2014
- Se agrega el módulo de instalación de
  SAHARA y las respectivas adecuaciones
  en los módulos de keystone, base de datos
  y scripts de soporte del instalador para
  Sahara.
- Mejoras en el README principal.

0.0.2.beta.el7 - 13-Oct-2014
- Segundo beta basado en JUNO (RC1/B3)
- Se desactiva SELINUX por bugs encontrados
  en NOVA-API y cuando se usa como backend
  de Base de Datos PostgreSQL
- Mejoras en módulos de trove y swift

0.0.1.beta.el7 - 12-Oct-2014
- Primer beta basado para JUNO (RC1/B3)
- Se cambian las imágenes de Cirros por la
  versión 0.3.3
- Modificaciones generales para adaptar el
  instalador a JUNO


main-installer.sh
#!/bin/bash
#
# Instalador desatendido para Openstack Juno sobre CENTOS
# Reynaldo R. Martinez P.
# E-Mail: TigerLinux@Gmail.com
# Primera versión (Grizzly): Julio 18 del 2013
# Primera versión (Havana - Centos6): Octubre 17 del 2013
# Primera versión (Havana - Debian7): Octubre 30 del 2013
# Primera versión para Icehouse (centos): Abril 15 del 2014
# Primera versión para Icehouse (debian): Abril 19 del 2014
# Primera versión para Icehouse (ubuntu): Abril 23 del 2014
# Primera versión para Juno (centos): Octubre 12 del 2014
# Primera versión para Juno (debian): Octubre 26 del 2014
# Primera versión para Juno (ubuntu): Octubre 29 del 2014
#
# Script principal
# Versión 1.0.8.el7 "Shadow Cat"
# 28 de Febrero del 2015
#

PATH=$PATH:/bin:/sbin:/usr/bin:/usr/sbin:/usr/local/bin:/usr/local/sbin

case $1 in
"install")

	if [ -f ./configs/main-config.rc ]
	then
		source ./configs/main-config.rc
		mkdir -p /etc/openstack-control-script-config
		date > /etc/openstack-control-script-config/install-init-date-and-time
		# Nuevo - Fix de permisologías y modos
		chown -R root.root *
		find . -name "*" -type f -exec chmod 644 "{}" ";"
		find . -name "*.sh" -type f -exec chmod 755 "{}" ";"
	else
		echo "No puedo acceder a mi archivo de configuración"
		echo "Revise que esté ejecutando el instalador en su directorio"
		echo "Abortando !!!!."
		echo ""
		exit 0
	fi

	clear

	echo ""
	echo "INSTALADOR DE OPENSTACK JUNO PARA CENTOS/RHEL/SCIENTIFIC-LINUX 7"
	echo "Para releases 7.0 en adelante"
	echo "Realizado por Reynaldo R. Martinez P."
	echo "E-Mail: TigerLinux@Gmail.com"
	echo "Versión 1.0.8.el7 \"Shadow Cat\" - Febrero 28, 2015"
	echo ""
	echo "Se verificaran los prerequisitos"
	echo "Si alguno de los prerequisitos falla, se informará y se detendrá el proceso"
	echo ""
	echo "Prerequisitos:"
	echo "- OS: RHEL 7.0 en adelante (o compatible): SL 7.0, Centos 7.0, etc. de 64 bits"
	echo "- El usuario que ejecuta este script debe ser root"
	echo "- El servidor debe tener los repositorios EPEL y RDO instalados:"
	echo "  * epel: http://fedoraproject.org/wiki/EPEL"
	echo "  * rdo:  http://repos.fedorapeople.org/repos/openstack/openstack-icehouse/epel-7/"
	echo "- Todos los repositorios deben estar disponibles para instalar los paquetes via yum"
	echo "- El servidor debe tener todas las actualizaciones y debe tener el Kernel de RDO instalado"
	echo "- OpenVSWITCH debe estar instalado y configurado con los bridges que se van a utilizar"
	echo "- Si desea instalar swift, los filesystems para el módulo swift deben estar montados en el"
	echo "  directorio /srv/node"
	echo ""
	echo "NOTA: Si quiere almacenar un log de todo lo que hace este instalador, use el comando tee."
	echo "Ejemplo: ./main-installer.sh install | tee -a /var/log/my_log_de_install.log"
	echo ""

	case $2 in
	auto|AUTO)
		echo "Modo automático activado"
		;;
	*)
		echo -n "Desea continuar ? [y/n]:"
		read -n 1 answer
		echo ""
		case $answer in
		y|Y)
			echo ""
			echo "Empezando Verificaciones Iniciales"
			echo ""
			;;
		*)
			echo ""
			echo "Abortando a petición del usuario/admin !!!"
			echo ""
			exit 0
			;;
		esac
		;;
	esac

	
	echo ""
	echo "Ejecutando validaciones iniciales e instalando requerimientos"
	echo ""

	./modules/requeriments.sh
	
	if [ -f /etc/openstack-control-script-config/libvirt-installed ]
	then
		echo ""
		echo "Requerimientos instalados"
		echo "Validaciones iniciales completadas"
		echo "Continuando con los módulos de la instalación"
		echo ""
	else
		echo ""
		echo "Falló la instalación de los requerimientos"
		echo "o fallaron las validaciones"
		echo "Abortando el resto de la instalación !!"
		echo ""
		exit 0
	fi


	rm -rf /tmp/keystone-signing-*
	rm -rf /tmp/cd_gen_*

	if [ $messagebrokerinstall == "yes" ]
	then
		echo ""
		echo "Instalando message broker"
		./modules/messagebrokerinstall.sh
		
		if [ -f /etc/openstack-control-script-config/broker-installed ]
		then
			echo ""
			echo "Listo"
			echo ""
		else
			echo ""
			echo "Falló la instalación del message broker"
			echo "Abortando el resto de la instalación"
			echo ""
			exit 0
		fi
	fi

	echo ""
	echo "Ejecutando módulo de soporte de base de datos"
	echo ""

	./modules/databaseinstall.sh

	if [ -f /etc/openstack-control-script-config/db-installed ]
	then
		echo ""
		echo "Módulo de soporte de bases de datos ejecutado con éxito"
		echo ""
	else
		echo ""
		echo "Falló el módulo de soporte de bases de datos"
		echo "Abortando el resto de la instalación"
		echo ""
		exit 0
	fi


	if [ $keystoneinstall == "yes" ]
	then
		echo ""
		echo "Instalando Keystone"

		./modules/keystoneinstall.sh

		if [ -f /etc/openstack-control-script-config/keystone-installed ]
		then
			echo "Keystone exitosamente instalado"
		else
			echo ""
			echo "Falló el módulo de instalación de keystone"
			echo "Abortando el resto de la instalación"
			echo ""
			exit 0
		fi

	else
		SERVICE_ENDPOINT="http://$keystonehost:35357/v2.0"
		OS_USERNAME=$keystoneadminuser
		OS_TENANT_NAME=$keystoneadminuser
		OS_PASSWORD=$keystoneadminpass
		# OS_AUTH_URL="http://$keystonehost:35357/v2.0/"
		OS_AUTH_URL="http://$keystonehost:5000/v2.0/"

		echo "# export SERVICE_ENDPOINT=$SERVICE_ENDPOINT" > $keystone_admin_rc_file
		echo "# export SERVICE_TOKEN=$SERVICE_TOKEN" >> $keystone_admin_rc_file
		echo "# export OS_SERVICE_TOKEN=$SERVICE_TOKEN" >> $keystone_admin_rc_file
		echo "export OS_USERNAME=$OS_USERNAME" >> $keystone_admin_rc_file
		echo "export OS_PASSWORD=$OS_PASSWORD" >> $keystone_admin_rc_file
		echo "export OS_TENANT_NAME=$OS_TENANT_NAME" >> $keystone_admin_rc_file
		echo "export OS_AUTH_URL=$OS_AUTH_URL" >> $keystone_admin_rc_file
		echo "PS1='[\u@\h \W(keystone_admin)]\$ '" >> $keystone_admin_rc_file

		mkdir -p /etc/openstack-control-script-config
		date > /etc/openstack-control-script-config/keystone-installed
		date > /etc/openstack-control-script-config/keystone-extra-idents
	fi

	if [ $swiftinstall == "yes" ]
	then
		echo ""
		echo "Instalando Swift"

		./modules/swiftinstall.sh

		if [ -f /etc/openstack-control-script-config/swift-installed ]
		then
			echo "Swift exitosamente instalado"
		else
			echo ""
			echo "Falló el módulo de instalación de swift"
			echo "Abortando el resto de la instalación"
			echo ""
			exit 0
		fi
	fi

	if [ $glanceinstall == "yes" ]
	then
		echo ""
		echo "Instalando Glance"

		./modules/glanceinstall.sh

		if [ -f /etc/openstack-control-script-config/glance-installed ]
		then
			echo "Glance exitosamente instalado"
		else
			echo ""
			echo "Falló el módulo de instalación de glance"
			echo "Abortando el resto de la instalación"
			echo ""
			exit 0
		fi
	fi

	if [ $cinderinstall == "yes" ]
	then
		echo ""
		echo "Instalando Cinder"

		./modules/cinderinstall.sh

		if [ -f /etc/openstack-control-script-config/cinder-installed ]
		then
			echo "Cinder exitosamente instalado"
		else
			echo ""
			echo "Falló el módulo de instalación de cinder"
			echo "Abortando el resto de la instalación"
			echo ""
			exit 0
		fi
	fi

	if [ $neutroninstall == "yes" ]
	then
		echo ""
		echo "Instalando Quantum"

		./modules/neutroninstall.sh

		if [ -f /etc/openstack-control-script-config/neutron-installed ]
		then
			echo "Neutron exitosamente instalado"
		else
			echo ""
			echo "Falló el módulo de instalación de neutron"
			echo "Abortando el resto de la instalación"
			echo ""
			exit 0
		fi
	fi

	if [ $novainstall == "yes" ]
	then
		echo ""
		echo "Instalando Nova"

		./modules/novainstall.sh

		if [ -f /etc/openstack-control-script-config/nova-installed ]
		then
			echo "Nova exitosamente instalado"
		else
			echo ""
			echo "Falló el módulo de instalación de nova"
			echo "Abortando el resto de la instalación"
			echo ""
			exit 0
		fi
	fi

	if [ $ceilometerinstall == "yes" ]
	then
		echo ""
		echo "Instalando Ceilometer"

		./modules/ceilometerinstall.sh

		if [ -f /etc/openstack-control-script-config/ceilometer-installed ]
		then
			echo "Ceilometer exitosamente instalado"
		else
			echo ""
			echo "Falló el módulo de instalación de ceilometer"
			echo "Abortando el resto de la instalación"
			echo ""
			exit 0
		fi
	fi

        if [ $heatinstall == "yes" ]
        then
                echo ""
                echo "Instalando Heat"

                ./modules/heatinstall.sh

                if [ -f /etc/openstack-control-script-config/heat-installed ]
                then
                        echo "Heat exitosamente instalado"
                else
                        echo ""
                        echo "Falló el módulo de instalación de heat"
                        echo "Abortando el resto de la instalación"
                        echo ""
                        exit 0
                fi
        fi

        case $dbflavor in
        "mysql")
                if [ $troveinstall == "yes" ]
                then
                        echo ""
                        echo "Instalando trove"

                        ./modules/troveinstall.sh

                        if [ -f /etc/openstack-control-script-config/trove-installed ]
                        then
                                echo "trove exitosamente instalado"
                        else
                                echo ""
                                echo "Falló el módulo de instalación de trove"
                                echo "Abortando el resto de la instalación"
                                echo ""
                                exit 0
                        fi
                fi
                ;;
        "postgres")
                if [ $troveinstall == "yes" ]
                then
                        echo ""
                        echo "Instalando trove"

                        ./modules/troveinstall.sh

                        if [ -f /etc/openstack-control-script-config/trove-installed ]
                        then
                                echo "trove exitosamente instalado"
                        else
                                echo ""
                                echo "Falló el módulo de instalación de trove"
                                echo "Abortando el resto de la instalación"
                                echo ""
                                exit 0
                        fi

                fi
                ;;
        esac

	if [ $saharainstall == "yes" ]
	then
		echo ""
		echo "Instalando sahara"

		./modules/saharainstall.sh

		if [ -f /etc/openstack-control-script-config/sahara-installed ]
		then
			echo "sahara exitosamente instalado"
		else
			echo ""
			echo "Falló el módulo de instalación de sahara"
			echo "Abortando el resto de la instalación"
			echo ""
			exit 0
		fi
	fi

	if [ $snmpinstall == "yes" ]
	then
		echo ""
		echo "Instalando infraestructura de monitoreo"

		./modules/snmpinstall.sh

		if [ -f /etc/openstack-control-script-config/snmp-installed ]
		then
			echo "Soporte SNMP exitosamente instalado"
		else
			echo ""
			echo "Falló la instalación del soporte SNMP, pero esto"
			echo "no es crítico - continuando con la instalación"
			echo ""
		fi
	fi

	if [ $horizoninstall == "yes" ]
	then
		echo ""
		echo "Instalando Dashboard Horizon"

		./modules/horizoninstall.sh

		if [ -f /etc/openstack-control-script-config/horizon-installed ]
		then
			echo "Horizon exitosamente instalado"
		else
			echo ""
			echo "Falló el módulo de instalación de horizon"
			echo "Abortando el resto de la instalación"
			echo ""
			exit 0
		fi

	fi

	echo ""
	echo "Ejecutando Post Install"
	./modules/postinstall.sh

	date > /etc/openstack-control-script-config/install-end-date-and-time

	echo ""
	echo "Instalación de OpenStack Finalizada exitosamente"
	echo ""
	
	;;
"uninstall")

	if [ -f ./configs/main-config.rc ]
	then
		source ./configs/main-config.rc
	else
		echo "No puedo acceder a mi archivo de configuración"
		echo "Revise que esté ejecutando el instalador en su directorio"
		echo "Abortando !!!!."
		echo ""
		exit 0
	fi

	echo ""
	echo "Se desinstalarán todos los Servicios de OpenStack"
	echo ""
	case $2 in
	auto|AUTO)
		echo "Modo automático activado"
		;;
	*)
		echo -n "Esta seguro que desea continuar ? [y/n]:"
		read -n 1 answer
		case $answer in
		y|Y)
			echo ""
			echo "Desinstalando Servicios de OpenStack"
			echo ""
			;;
		*)
			echo ""
			echo "Se abortará la desinstalación a petición del usuario/admin"
			echo ""
			exit 0
			;;
		esac
	esac
	./modules/uninstall.sh
	;;
*)
	echo ""
	echo "Uso: $0 install | uninstall [auto]"
	echo ""
	;;
esac

